technic/machines/other/injector.lua

```diff
@@ -112,46 +70,96 @@ minetest.register_node("technic:injector", {
-		if fields.mode_stack then meta:set_string("mode", "whole stacks") end
-		if fields["fs_helpers_cycling:0:splitstacks"]
-		  or fields["fs_helpers_cycling:1:splitstacks"] then
-			fs_helpers.on_receive_fields(pos, fields)
+		if fields.stackwise then
+			meta:set_int("stackwise", 1)
+		elseif fields.itemwise then
+			meta:set_int("stackwise", 0)
+		elseif fields.disable then
+			meta:set_int("disabled", 1)
+			minetest.get_node_timer(pos):stop()
+		elseif fields.enable then
+			meta:set_int("disabled", 0)
+			minetest.get_node_timer(pos):start(1)
 		end
+		fs_helpers.on_receive_fields(pos, fields)
 		set_injector_formspec(meta)
 	end,
+	on_timer = function(pos, elapsed)
+		local meta = minetest.get_meta(pos)
+		if meta:get_int("disabled") == 1 then
+			return false
+		end
+		local node = minetest.get_node(pos)
+		local dir = param2_to_under[math.floor(node.param2 / 4)]
+		local node_under = minetest.get_node(vector.add(pos, dir))
+		if minetest.get_item_group(node_under.name, "tubedevice") > 0 then
+			local inv = meta:get_inventory()
+			local list = inv:get_list("main")
+			if not list then
+				return true
+			end
+			local stackwise = meta:get_int("stackwise") == 1
+			for i,stack in ipairs(list) do
+				if not stack:is_empty() then
+					if stackwise then
+						technic.tube_inject_item(pos, pos, dir, stack:to_table())
+						stack:clear()
+					else
+						technic.tube_inject_item(pos, pos, dir, stack:take_item(1):to_table())
+					end
+					inv:set_stack("main", i, stack)
+					break
+				end
+			end
+		end
+		return true
+	end,
+	on_rotate = function(pos, node, user, mode, new_param2)
+		node.param2 = new_param2
+		minetest.swap_node(pos, node)
+		pipeworks.scan_for_tube_objects(pos)
+		return true
+	end,
 	allow_metadata_inventory_put = technic.machine_inventory_put,
 	allow_metadata_inventory_take = technic.machine_inventory_take,
 	allow_metadata_inventory_move = technic.machine_inventory_move,
 	after_place_node = pipeworks.after_place,
 	after_dig_node = pipeworks.after_dig
 })
 
-minetest.register_abm({
-	label = "Machines: run injector",
+minetest.register_craft({
+	output = "technic:injector 1",
+	recipe = {
+		{"", "technic:control_logic_unit",""},
+		{"", "default:chest",""},
+		{"", "pipeworks:tube_1",""},
+	}
+})
+
+minetest.register_lbm({
+	label = "Old injector conversion",
+	name = "technic:old_injector_conversion",
 	nodenames = {"technic:injector"},
-	interval = 1,
-	chance = 1,
-	action = function(pos, node, active_object_count, active_object_count_wider)
-		local dir = param2_to_under[math.floor(node.param2/4)]
-		local pos1 = vector.add(pos, dir)
-		local node1 = minetest.get_node(pos1)
-		if minetest.get_item_group(node1.name, "tubedevice") > 0 then
-			inject_items(pos, dir)
+	run_at_every_load = false,
+	action = function(pos, node)
+		local meta = minetest.get_meta(pos)
+		if meta:get_string("mode") == "whole stacks" then
+			meta:set_int("stackwise", 1)
```

Comment:
Is there good enough reason to change meta field names and value type?
Keeping it how it originally was would allow bit better compatibility, allows rolling back without changing code.